File: src/main/java/rars/util/BitmapDisplay.java
package rars.util;

import org.apache.logging.log4j.LogManager;
import org.apache.logging.log4j.Logger;
import org.jetbrains.annotations.NotNull;
import org.jetbrains.annotations.Nullable;
import rars.assembler.DataTypes;
import rars.exceptions.AddressErrorException;
import rars.notices.AccessNotice;
import rars.notices.MemoryAccessNotice;
import rars.riscv.hardware.Memory;
import rars.venus.VenusUI;

import javax.swing.*;

public final class BitmapDisplay extends JFrame {

    private static final @NotNull Logger LOGGER = LogManager.getLogger(BitmapDisplay.class);

    public final int displayWidth;
    public final int displayHeight;
    private final @NotNull GraphicsPanel panel;
    private @Nullable ListenerDispatcher<@NotNull MemoryAccessNotice>.Handle listenerHandle;
    private final @NotNull Memory memory;
    private int baseAddress;
    private int upperAddressBound;

    public BitmapDisplay(
        final @NotNull Memory memory,
        final @NotNull VenusUI mainUI,
        final int baseAddress,
        final int displayWidth,
        final int displayHeight
    ) {
        super();
        this.memory = memory;
        this.baseAddress = baseAddress;
        this.displayWidth = displayWidth;
        this.displayHeight = displayHeight;
        this.upperAddressBound = baseAddress + (displayWidth * displayHeight * DataTypes.WORD_SIZE);

        this.setTitle("Syscall: DisplayBitmap");
        this.setDefaultCloseOperation(JFrame.DISPOSE_ON_CLOSE);
        this.panel = new GraphicsPanel(displayWidth, displayHeight);
        
        this.add(this.panel);
        this.setResizable(false);
        this.setLocationRelativeTo(mainUI);

        this.fillGrid();
        this.pack();

        try {
            this.listenerHandle = this.memory.subscribe(
                this::onMemoryAccess, baseAddress, upperAddressBound
            );
        } catch (final AddressErrorException e) {
            throw new RuntimeException(e);
        }
    }

    private void onMemoryAccess(final @NotNull MemoryAccessNotice notice) {
        if (notice.accessType == AccessNotice.AccessType.WRITE) {
            this.updateDisplay(notice.address, notice.length);
        }
    }

    public void changeBaseAddress(final int newBaseAddress) {
        this.unsubscribeFromMemory();
        this.baseAddress = newBaseAddress;
        this.upperAddressBound = newBaseAddress + (this.displayWidth * this.displayHeight * DataTypes.WORD_SIZE);
        try {
            this.listenerHandle = this.memory.subscribe(
                this::onMemoryAccess,
                this.baseAddress,
                this.upperAddressBound
            );
        } catch (final AddressErrorException e) {
            throw new RuntimeException(e);
        }
    }

    public void unsubscribeFromMemory() {
        if (this.listenerHandle != null) {
            this.memory.deleteSubscriber(this.listenerHandle);
        }
        this.listenerHandle = null;
    }
    
    public int getBaseAddress() {
        return this.baseAddress;
    }

    private void fillGrid() {
        var currentOffset = 0;
        for (int row = 0; row < this.displayHeight; row++) {
            for (int col = 0; col < this.displayWidth; col++) {
                final int address = this.baseAddress + currentOffset;
                try {
                    final var word = this.memory.getWordNoNotify(address);
                    this.panel.paintPixel(row, col, word);
                } catch (final AddressErrorException e) {
                    LOGGER.error("Error updating color for address {} in bitmap display: {}", address, e);
                    return;
                }
                currentOffset += DataTypes.WORD_SIZE;
            }
        }
    }

    private void updateDisplay(final int memoryAddress, final int writeLength) {

        // figure out which pixels were changed
        final int endAddress = memoryAddress + writeLength;
        // clamp the range to the display bounds
        if (endAddress >= this.baseAddress && memoryAddress <= this.upperAddressBound) {
            // the memory written may not be aligned by 4 bytes, so we round
            // the start and end to the nearest 4-byte boundary
            final var start = (Math.max(memoryAddress, this.baseAddress) / 4) * 4;
            final var end = ((Math.min(endAddress, this.upperAddressBound) + 3) / 4) * 4;
            // these values are already nicely aligned, so all that's left to
            // do is to update the grid
            var row = (start - this.baseAddress) / (this.displayWidth * 4);
            var col = (start - this.baseAddress) % (this.displayWidth * 4) / 4;
            for (int i = start; i < end; i += 4) {
                try {
                    final var word = this.memory.getWordNoNotify(i);
                    this.panel.paintPixel(row, col, word);
                } catch (final AddressErrorException e) {
                    throw new RuntimeException(e);
                }
                col++;
                if (col == this.displayWidth) {
                    col = 0;
                    row++;
                }
            }
        }
    }
}

File: src/main/java/rars/util/BitmapDisplay.kt
package rars.util

import arrow.core.Either
import rars.assembler.DataTypes
import rars.logging.RARSLogging
import rars.logging.error
import rars.notices.AccessType
import rars.notices.MemoryAccessNotice
import rars.riscv.hardware.memory.Memory
import rars.riscv.hardware.memory.MemoryListenerHandle
import java.awt.Color
import java.awt.Dimension
import javax.swing.JFrame
import kotlin.math.max
import kotlin.math.min

// TODO: move this class to the common SystemIO class
class BitmapDisplay(
    private val memory: Memory,
    @JvmField var baseAddress: Int,
    @JvmField val displayWidth: Int,
    @JvmField val displayHeight: Int
) : JFrame() {
    private var upperAddressBound: Int =
        baseAddress + (displayWidth * displayHeight * DataTypes.WORD_SIZE)
    private val grid: Grid = Grid(displayHeight, displayWidth)
    private val panel: GraphicsPanel =
        GraphicsPanel(Dimension(displayWidth, displayHeight), this.grid)
    private val accessNoticeCallback: (MemoryAccessNotice) -> Unit = { notice ->
        if (notice.accessType == AccessType.WRITE) {
            this.updateDisplay(notice.address, notice.length)
        }
    }
    private var handle: MemoryListenerHandle<Int>

    init {
        title = "Syscall: DisplayBitmap"
        defaultCloseOperation = DISPOSE_ON_CLOSE
        add(panel)
        isResizable = false

        fillGrid()
        pack()

        handle = memory
            .subscribe(accessNoticeCallback, baseAddress, upperAddressBound)
            .unwrap()
    }

    fun changeBaseAddress(newBaseAddress: Int) {
        memory.unsubscribe(handle)
        baseAddress = newBaseAddress
        upperAddressBound =
            newBaseAddress + (displayWidth * displayHeight * DataTypes.WORD_SIZE)
        handle = memory.subscribe(
            accessNoticeCallback,
            baseAddress,
            upperAddressBound
        ).unwrap()
    }

    fun unsubscribeFromMemory() {
        memory.unsubscribe(handle)
    }

    private fun fillGrid() {
        var currentOffset = 0
        for (row in 0..<displayHeight) {
            for (col in 0..<displayWidth) {
                val address = baseAddress + currentOffset
                val word =
                    memory.silentMemoryView.getWord(address).unwrap { e ->
                        LOGGER.error {
                            """Error updating color for address $address in bitmap display: $e"""
                        }
                        return
                    }
                val color = Color(word)
                grid.setColor(row, col, color)
                currentOffset += DataTypes.WORD_SIZE
            }
        }
    }

    private fun updateDisplay(memoryAddress: Int, writeLength: Int) {
        // figure out which pixels were changed

        val endAddress = memoryAddress + writeLength
        // clamp the range to the display bounds
        if (endAddress >= baseAddress && memoryAddress <= upperAddressBound) {
            // the memory written may not be aligned by 4 bytes, so we round
            // the start and end to the nearest 4 byte boundary
            val start = (max(memoryAddress, baseAddress) / 4) * 4
            val end = ((min(endAddress, upperAddressBound) + 3) / 4) * 4
            // these values are already nicely aligned, so all that's left to
            // do is to update the grid
            var row = (start - baseAddress) / (displayWidth * 4)
            var col = (start - baseAddress) % (displayWidth * 4) / 4
            var i = start
            while (i < end) {
                when (val word = memory.silentMemoryView.getWord(i)) {
                    is Either.Left -> {
                        LOGGER.error {
                            "Error updating color for address $i in bitmap display: ${word.value}"
                        }
                        break
                    }
                    is Either.Right -> {
                        val color = Color(word.value)
                        grid.setColor(row, col, color)
                        col++
                        if (col == displayWidth) {
                            col = 0
                            row++
                        }
                        i += 4
                    }
                }
            }
        }
        panel.repaint()
    }

    companion object {
        private val LOGGER = RARSLogging.forClass(BitmapDisplay::class)
    }
}
